src/payment/strategies/adyenv2/adyenv2-payment-strategy.ts

```diff
@@ -143,66 +141,37 @@ export default class AdyenV2PaymentStrategy implements PaymentStrategy {
                 return this._store.dispatch(this._paymentActionCreator.submitPayment(paymentPayload));
             })
             .catch(error => {
-                if (!(error instanceof RequestError) || (!some(error.body.errors, { code: 'three_d_secure_required' }) && !some(error.body.errors, { code: 'additional_action_required' }))) {
+                if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'additional_action_required' })) {
                     return Promise.reject(error);
                 }
 
-                if (error.body.three_ds_result.code === ResultCode.IdentifyShopper) {
-                    return this._handle3DS2Fingerprint(error.body.three_ds_result, payment.methodId)
-                        .then((payment: Payment) =>
-                            this._store.dispatch(this._paymentActionCreator.submitPayment({
-                                ...payment,
-                                paymentData: {
-                                    ...payment.paymentData,
-                                    shouldSaveInstrument,
-                                },
-                            })))
-                        .catch(error => {
-                            if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'three_d_secure_required' })) {
-                                return Promise.reject(error);
-                            }
-
-                            if (error.body.three_ds_result.code === ResultCode.ChallengeShopper) {
-                                return this._handle3DS2Challenge(error.body.three_ds_result, payment.methodId)
-                                    .then((payment: Payment) =>
-                                        this._store.dispatch(this._paymentActionCreator.submitPayment({
-                                            ...payment,
-                                            paymentData: {
-                                                ...payment.paymentData,
-                                                shouldSaveInstrument,
-                                            },
-                                        }))
-                                    );
-                            }
-
+                return this._handleFromAction(error.body.provider_data)
+                    .then((payment: Payment) =>
+                        this._store.dispatch(this._paymentActionCreator.submitPayment({
+                            ...payment,
+                            paymentData: {
+                                ...payment.paymentData,
+                                shouldSaveInstrument,
+                            },
+                        }))
+                    )
+                    .catch(error => {
```

Comment:
You might want to extract this catch block into a private method, it seems to be repeated one after another.

```ts
.catch(error => {
    if (!(error instanceof RequestError) || !some(error.body.errors, { code: 'additional_action_required' })) {
        return Promise.reject(error);
    }

    return this._handleFromAction(error.body.provider_data)
        .then((payment: Payment) =>
            this._store.dispatch(this._paymentActionCreator.submitPayment({
                ...payment,
                paymentData: {
                    ...payment.paymentData,
                    shouldSaveInstrument,
                },
            }))
        );
})
```

A second question is, do we really need to have two `catch` blocks? As in, is it possible for a provider to ask for two additional actions?

Reply:
At least there's one case where two additional actions can be fired and it's when [Challenge](https://docs.adyen.com/checkout/3d-secure/native-3ds2/web-component) flow can be preceded by a frictionless flow.